lib/elixir/pages/Guards.md

```diff
@@ -0,0 +1,226 @@
+    * `is_binary/1`
+    * `is_bitstring/1`
+    * `is_boolean/1`
+    * `is_float/1`
+    * `is_function/1`
+    * `is_function/2`
+    * `is_integer/1`
+    * `is_list/1`
+    * `is_map/1`
+    * `is_nil/1`
+    * `is_number/1`
+    * `is_pid/1`
+    * `is_port/1`
+    * `is_reference/1`
+    * `is_tuple/1`
+  * the following guard-friendly functions (all documented in the `Kernel` module):
+    * `abs/1`
+    * `binary_part/3`
+    * `bit_size/1`
+    * `byte_size/1`
+    * `div/2`
+    * `elem/2`
+    * `hd/1`
+    * `length/1`
+    * `map_size/1`
+    * `node/0`
+    * `node/1`
+    * `rem/2`
+    * `round/1`
+    * `self/0`
+    * `tl/1`
+    * `trunc/1`
+    * `tuple_size/1`
+  * the following handful of Erlang bitwise operations, if imported from the `Bitwise` module:
+    * `band/2` or the `&&&` operator
+    * `bor/2` or the `|||` operator
+    * `bnot/1` or the `~~~` operator
+    * `bsl/1` or the `<<<` operator
+    * `bsr/1` or the `>>>` operator
+    * `bxor/2` or the `^^^` operator
+
+## Defining custom guard expressions
+
+As mentioned before, only the expressions listed in the previous section are allowed in guards. However, we can take advantage of macros to write custom guards that can simplify our programs or make them more domain-specific. At the end of the day, what matters is that the *output* of the macros (which is what will be compiled) boils down to a combinations of the allowed expressions.
+
+Let's look at a quick case study: we want to check that a function argument is a keyword list (a list of zero or more `{key, value}` pairs where `key` is an atom). With pattern matching, this is impossible to do since the list could be of any length. It's impossible to check this with guards as well: we can't check *all* of the elements of the list; however, we can perform a weak check by checking that only the first element of the list is a `{atom, value}` tuple. While this does not ensure the argument is a keyword list, it may be enough to be useful in real-world scenarios.
+
+This guard would look like this:
+
+```elixir
+def is_keyword_list(list)
+    when is_list(list) and
+         is_tuple(hd(list)) and
+         tuple_size(hd(list)) == 2 and
+         is_atom(elem(hd(list), 0)) do
+  true
+end
+
+def keyword_list?(_) do
+  false
+end
+```
+
+This would be terribly complex to write everytime we need this check, so, as mentioned at the beginning of this section, we can abstract this away using a macro. Remember that defining a function that performs this check wouldn't work because we can't use custom functions in guards. Our macro would look like this:
+
+```elixir
+defmacro is_keyword_list(list) do
+  quote do
+    is_list(unquote(list)) and
+      is_tuple(hd(unquote(list))) and
+      tuple_size(hd(unquote(list))) == 2 and
+      is_atom(elem(hd(unquote(list)), 0))
+  end
+end
+```
+
+It's important to note that `list` needs to be unquoted every time it is used, as it's not possible to do something like this:
```

Comment:
Yes, but this is bad advice if you eventually use `is_keyword_list` outside of a guard. :)